### PR TITLE
[components][driver]add phy and mdio bus

### PR DESCRIPTION
**Pull request by zhuzhuzhus**
## 拉取/合并请求描述：(PR description)

[

#### 为什么提交这份PR (why to submit this PR)
old phy bus is too simple add phy_bus is not adapt rt_bus framework,so writer a stronger phy bus framework.
 
here is my commit message:
add mdio bus and phy bus to kernel,the phy bus use rt_bus framewok ,driver writer can write phy_driver first .when mac driver need to use phy they can register phy_device and pjhy_devie will serach for driver which match by uid and mask,if no driver match with the device that you register,phy_bus will return the genphy to you device,the genphy driver is the general driver for phy,so you can use it but it can not support the capcity of chip it may be cause performance is not up to peak

#### 请提供验证的bsp和config (provide the config and bsp) 

- BSP: rk3568

- .config:

- action:

]

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other styles
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 代码是高质量的 Code in this PR is of high quality
- [x] 已经使用[formatting](https://github.com/mysterywolf/formatting) 等源码格式化工具确保格式符合[RT-Thread代码规范](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_cn.md) This PR complies with [RT-Thread code specification](https://github.com/RT-Thread/rt-thread/blob/master/documentation/contribution_guide/coding_style_en.md)

### CODE REVIEW

**Comment by zhuzhuzhus** [comment]
https://github.com/RT-Thread/rt-thread/pull/9470#issue-2547447682
依赖于该pr

File: components/drivers/phy/phy.c [hunk 1]
```diff
@@ -1,76 +1,514 @@
...
+    dev->interface = RT_PHY_INTERFACE_MODE_NA;
+#ifdef RT_USING_OFW
+    dev->node = RT_NULL;
 #endif
+    dev->autoneg = RT_TRUE;
 
-/*
-* phy device register
-*/
-rt_err_t rt_hw_phy_register(struct rt_phy_device *phy, const char *name)
```
[the remark this answers is not in the document]
Comment (zhuzhuzhus) [reply components/drivers/phy/phy.c:49]: 没想到有bsp用到了这玩意，那我看看怎么修改

File: components/drivers/phy/general_phy.h [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+#define RT_EXPANSION_REG_NPCAPABLE  0x0008  /* Link partner supports npage */
+#define RT_EXPANSION_REG_MFAULTS    0x0010  /* Multiple faults detected    */
+#define RT_EXPANSION_REG_RESV       0xffe0  /* Unused...                   */
+
+#define RT_SUPORT_1000B_XFULL   0x8000  /* Can do 1000BX Full */
+#define RT_SUPORT_1000B_XHALF   0x4000  /* Can do 1000BX Half */
+#define RT_SUPORT_1000B_TFULL   0x2000  /* Can do 1000BT Full          */
+#define RT_SUPORT_1000B_THALF   0x1000  /* Can do 1000BT Half          */
+#define RT_PHY_ANEG_TIMEOUT    4000
+struct rt_phy_device;
```
Comment (zmshahaha) [inline components/drivers/phy/general_phy.h:136]: ?

File: components/drivers/phy/ofw.c [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+        if (len >= strlen("ethernet-phy-idVVVV.DDDD"))
+        {
+            char *s = strstr(phy_id, "ethernet-phy-id");
+
+            /*
+             * check if the string is something like
+             * ethernet-phy-idVVVV.DDDD
+             */
+            if (s && s[19] == '.')
```
Comment (zmshahaha) [inline components/drivers/phy/ofw.c:129]: 19这个数字咋来的，最好别出现纯数字

Comment (zhuzhuzhus) [reply components/drivers/phy/ofw.c:129]: image
phy节点固定的，这样没什么问题，也不会变

**Comment by zmshahaha** [comment]
不要乱标resolved

File: components/drivers/phy/ofw.c (hunk 3 above)
Comment (zmshahaha) [reply components/drivers/phy/ofw.c:129]: 这个还没有resolved

**Comment by zhuzhuzhus** [comment]
> 请不要随意标resolved，上面问题还没解决

哪个

File: components/drivers/phy/ofw.c (hunk 3 above)
Comment (zmshahaha) [reply components/drivers/phy/ofw.c:129]: 改成sizeof("ethernet-phy.....")吧

**Review by zmshahaha — CHANGES_REQUESTED** [review]

File: components/drivers/phy/ofw.c (hunk 3 above)
Comment (zmshahaha) [reply components/drivers/phy/ofw.c:129]: 换成sizeof("ether.....")吧

Comment (zhuzhuzhus) [reply components/drivers/phy/ofw.c:129]: 这个感觉没必要吧，用19反而更直观？

Comment (zmshahaha) [reply components/drivers/phy/ofw.c:129]: > s[sizeof(ethernet-phy-idVVVV.)/sizeof(s[0])]

?

Comment (zmshahaha) [reply components/drivers/phy/ofw.c:129]: s[sizeof(ethernet-phy-idVVVV.)]

Comment (zhuzhuzhus) [reply components/drivers/phy/ofw.c]: 我改了一种配对方式，用sscanf了

Comment (zmshahaha) [reply components/drivers/phy/ofw.c]: 🆗

**Comment by Rbb666** [comment]
@zhuzhuzhus PR有冲突了，帮忙解决下

**Comment by zhuzhuzhus** [comment]
> @zhuzhuzhus PR有冲突了，帮忙解决下

ok啦

File: components/drivers/phy/phy.c [hunk 4]
```diff
@@ -1,22 +1,22 @@
...
+ * 2020-09-27     wangqiang   first version
+ * 2024-10-08     zhujiale    add phy v2.0
  */
 #include <stddef.h>
 #include <rthw.h>
-#include <rtthread.h>
-#include <rtdevice.h>
-
-#define DBG_TAG "PHY"
+#include <drivers/phy.h>
```
Comment (Rbb666) [inline components/drivers/phy/phy.c:13]: 13、14行统一 #include rtdevice.h 就行

File: components/drivers/phy/ofw.c [hunk 5]
```diff
@@ -8,6 +8,11 @@
  * 2024-09-25     zhujiale    the first version
  */
 #include <rtthread.h>
+#include <drivers/phy.h>
```
Comment (Rbb666) [inline components/drivers/phy/ofw.c:11]: 这块也是，引用rtdevice.h就行

File: components/drivers/phy/mdio.h [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+ * Change Logs:
+ * Date           Author       Notes
+ * 2024-10-08     zhujiale    the first version
+ */
+#ifndef __PHY_MDIO_H__
+#define __PHY_MDIO_H__
+#include <rtthread.h>
+#include <drivers/phy.h>
+
+#define RT_MDIO_DEVAD_NONE          (-1)
```
Comment (Rbb666) [inline components/drivers/phy/mdio.h:15]: 注释看着乱乱的，可以格式化下吗？

File: components/drivers/phy/mdio.h [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Change Logs:
+ * Date           Author       Notes
+ * 2024-10-08     zhujiale    the first version
+ */
+#ifndef __PHY_MDIO_H__
+#define __PHY_MDIO_H__
+#include <rtthread.h>
+#include <drivers/phy.h>
```
Comment (Rbb666) [inline components/drivers/phy/mdio.h:13]: #include rtdevice.h

File: components/drivers/phy/general_phy.h [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Change Logs:
+ * Date           Author       Notes
+ * 2024-10-08     zhujiale    the first version
+ */
+#ifndef __PHY_GENERAL_H__
+#define __PHY_GENERAL_H__
+/* The forced speed, 10Mb, 100Mb, gigabit, 2.5Gb, 10GbE. */
+#define SPEED_10        10
```
Comment (Rbb666) [inline components/drivers/phy/general_phy.h:14]: 这个文件的注释也格式化下吧

**Review by Rbb666 — APPROVED** [review]

**Comment by zhuzhuzhus** [comment]
> @zhuzhuzhus 还有冲突，麻烦解下谢谢

我这里没有看到有冲突呀
image

**Pull request merged**